### PR TITLE
feat: shelf metrics should report on op-status for components

### PR DESCRIPTION
**Pull request by Hardikl**
feat: Shelf metrics should report on op-status for components like fan, psu.

shelf_sensor_labels:

shelf_sensor_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower left power supply", sensor_id="1", shelf="1.0", status="normal"} | 1
-- | --
shelf_sensor_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower left power supply", sensor_id="1", shelf="1.10", status="normal"} | 1
shelf_sensor_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower left power supply", sensor_id="2", shelf="1.0", status="normal"} | 1
shelf_sensor_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower left power supply", sensor_id="2", shelf="1.10", status="normal"} | 1
shelf_sensor_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower right power supply", sensor_id="3", shelf="1.0", status="normal"} | 1
shelf_sensor_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower right power supply", sensor_id="3", shelf="1.10", status="normal"} | 1
shelf_sensor_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower right power supply", sensor_id="4", shelf="1.0", status="normal"} | 1
shelf_sensor_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower right power supply", sensor_id="4", shelf="1.10", status="normal"}

shelf_psu_labels:

shelf_psu_labels{cluster="F8080-32-25", datacenter="DC-03", enabled="true", instance="localhost:12991", job="prometheus1", location="rear of the shelf at the bottom left", part_number="114-00065+B0", psu_id="1", serial="PSD041150689535", shelf="1.10", status="normal", type="9C"} | 1
-- | --
shelf_psu_labels{cluster="F8080-32-25", datacenter="DC-03", enabled="true", instance="localhost:12991", job="prometheus1", location="rear of the shelf at the bottom left", part_number="114-00065+B0", psu_id="1", serial="PSD041151523314", shelf="1.0", status="normal", type="9C"} | 1
shelf_psu_labels{cluster="F8080-32-25", datacenter="DC-03", enabled="true", instance="localhost:12991", job="prometheus1", location="rear of the shelf at the bottom right", part_number="114-00065+B0", psu_id="2", serial="PSD041150689534", shelf="1.10", status="normal", type="9C"} | 1
shelf_psu_labels{cluster="F8080-32-25", datacenter="DC-03", enabled="true", instance="localhost:12991", job="prometheus1", location="rear of the shelf at the bottom right", part_number="114-00065+B0", psu_id="2", serial="PSD041151523313", shelf="1.0", status="normal", type="9C"}

shelf_fan_labels:

shelf_fan_labels{cluster="F8080-32-25", datacenter="DC-03", fan_id="1", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower left power supply", shelf="1.0", status="normal"} | 1
-- | --
shelf_fan_labels{cluster="F8080-32-25", datacenter="DC-03", fan_id="1", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower left power supply", shelf="1.10", status="normal"} | 1
shelf_fan_labels{cluster="F8080-32-25", datacenter="DC-03", fan_id="2", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower left power supply", shelf="1.0", status="normal"} | 1
shelf_fan_labels{cluster="F8080-32-25", datacenter="DC-03", fan_id="2", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower left power supply", shelf="1.10", status="normal"} | 1
shelf_fan_labels{cluster="F8080-32-25", datacenter="DC-03", fan_id="3", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower right power supply", shelf="1.0", status="normal"} | 1
shelf_fan_labels{cluster="F8080-32-25", datacenter="DC-03", fan_id="3", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower right power supply", shelf="1.10", status="normal"} | 1
shelf_fan_labels{cluster="F8080-32-25", datacenter="DC-03", fan_id="4", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower right power supply", shelf="1.0", status="normal"} | 1
shelf_fan_labels{cluster="F8080-32-25", datacenter="DC-03", fan_id="4", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower right power supply", shelf="1.10", status="normal"}

shelf_temperature_labels:

shelf_temperature_labels{cluster="F8080-32-25", datacenter="DC-03", high_critical="42", high_warning="40", instance="localhost:12991", job="prometheus1", low_critical="0", low_warning="5", sensor_id="1", shelf="1.0", status="normal", temp_is_ambient="true"}
1
shelf_temperature_labels{cluster="F8080-32-25", datacenter="DC-03", high_critical="42", high_warning="40", instance="localhost:12991", job="prometheus1", low_critical="0", low_warning="5", sensor_id="1", shelf="1.10", status="normal", temp_is_ambient="true"}
1
shelf_temperature_labels{cluster="F8080-32-25", datacenter="DC-03", high_critical="55", high_warning="50", instance="localhost:12991", job="prometheus1", low_critical="5", low_warning="10", sensor_id="2", shelf="1.0", status="normal", temp_is_ambient="false"}
1
shelf_temperature_labels{cluster="F8080-32-25", datacenter="DC-03", high_critical="55", high_warning="50", instance="localhost:12991", job="prometheus1", low_critical="5", low_warning="10", sensor_id="2", shelf="1.10", status="normal", temp_is_ambient="false"}
1
shelf_temperature_labels{cluster="F8080-32-25", datacenter="DC-03", high_critical="55", high_warning="50", instance="localhost:12991", job="prometheus1", low_critical="5", low_warning="10", sensor_id="3", shelf="1.0", status="normal", temp_is_ambient="false"}
1
shelf_temperature_labels{cluster="F8080-32-25", datacenter="DC-03", high_critical="55", high_warning="50", instance="localhost:12991", job="prometheus1", low_critical="5", low_warning="10", sensor_id="3", shelf="1.10", status="normal", temp_is_ambient="false"}
1
shelf_temperature_labels{cluster="F8080-32-25", datacenter="DC-03", high_critical="55", high_warning="50", instance="localhost:12991", job="prometheus1", low_critical="5", low_warning="10", sensor_id="5", shelf="1.0", status="normal", temp_is_ambient="false"}
1
shelf_temperature_labels{cluster="F8080-32-25", datacenter="DC-03", high_critical="55", high_warning="50", instance="localhost:12991", job="prometheus1", low_critical="5", low_warning="10", sensor_id="5", shelf="1.10", status="normal", temp_is_ambient="false"}
1
shelf_temperature_labels{cluster="F8080-32-25", datacenter="DC-03", high_critical="60", high_warning="55", instance="localhost:12991", job="prometheus1", low_critical="5", low_warning="10", sensor_id="7", shelf="1.0", status="normal", temp_is_ambient="false"}
1
shelf_temperature_labels{cluster="F8080-32-25", datacenter="DC-03", high_critical="60", high_warning="55", instance="localhost:12991", job="prometheus1", low_critical="5", low_warning="10", sensor_id="7", shelf="1.10", status="normal", temp_is_ambient="false"}
1
shelf_temperature_labels{cluster="F8080-32-25", datacenter="DC-03", high_critical="60", high_warning="55", instance="localhost:12991", job="prometheus1", low_critical="5", low_warning="10", sensor_id="8", shelf="1.0", status="normal", temp_is_ambient="false"}
1
shelf_temperature_labels{cluster="F8080-32-25", datacenter="DC-03", high_critical="60", high_warning="55", instance="localhost:12991", job="prometheus1", low_critical="5", low_warning="10", sensor_id="8", shelf="1.10", status="normal", temp_is_ambient="false"}
1
shelf_temperature_labels{cluster="F8080-32-25", datacenter="DC-03", high_critical="70", high_warning="65", instance="localhost:12991", job="prometheus1", low_critical="5", low_warning="10", sensor_id="4", shelf="1.0", status="normal", temp_is_ambient="false"}
1
shelf_temperature_labels{cluster="F8080-32-25", datacenter="DC-03", high_critical="70", high_warning="65", instance="localhost:12991", job="prometheus1", low_critical="5", low_warning="10", sensor_id="4", shelf="1.10", status="normal", temp_is_ambient="false"}
1
shelf_temperature_labels{cluster="F8080-32-25", datacenter="DC-03", high_critical="70", high_warning="65", instance="localhost:12991", job="prometheus1", low_critical="5", low_warning="10", sensor_id="6", shelf="1.0", status="normal", temp_is_ambient="false"}
1
shelf_temperature_labels{cluster="F8080-32-25", datacenter="DC-03", high_critical="70", high_warning="65", instance="localhost:12991", job="prometheus1", low_critical="5", low_warning="10", sensor_id="6", shelf="1.10", status="normal", temp_is_ambient="false"}
1

shelf_voltage_labels:

shelf_voltage_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower left power supply", sensor_id="1", shelf="1.0", status="normal"} | 1
-- | --
shelf_voltage_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower left power supply", sensor_id="1", shelf="1.10", status="normal"} | 1
shelf_voltage_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower left power supply", sensor_id="2", shelf="1.0", status="normal"} | 1
shelf_voltage_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower left power supply", sensor_id="2", shelf="1.10", status="normal"} | 1
shelf_voltage_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower right power supply", sensor_id="3", shelf="1.0", status="normal"} | 1
shelf_voltage_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower right power supply", sensor_id="3", shelf="1.10", status="normal"} | 1
shelf_voltage_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower right power supply", sensor_id="4", shelf="1.0", status="normal"} | 1
shelf_voltage_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower right power supply", sensor_id="4", shelf="1.10", status="normal"} | 1
shelf_voltage_labels{cluster="fas8040-206-21", datacenter="DC-01", instance="localhost:12992", job="prometheus2", location="rear of the shelf on the lower left power supply", sensor_id="1", shelf="1.1", status="normal"} | 1
shelf_voltage_labels{cluster="fas8040-206-21", datacenter="DC-01", instance="localhost:12992", job="prometheus2", location="rear of the shelf on the lower left power supply", sensor_id="1", shelf="1.2", status="normal"} | 1
shelf_voltage_labels{cluster="fas8040-206-21", datacenter="DC-01", instance="localhost:12992", job="prometheus2", location="rear of the shelf on the lower left power supply", sensor_id="2", shelf="1.1", status="normal"} | 1
shelf_voltage_labels{cluster="fas8040-206-21", datacenter="DC-01", instance="localhost:12992", job="prometheus2", location="rear of the shelf on the lower left power supply", sensor_id="2", shelf="1.2", status="normal"} | 1
shelf_voltage_labels{cluster="fas8040-206-21", datacenter="DC-01", instance="localhost:12992", job="prometheus2", location="rear of the shelf on the lower right power supply", sensor_id="3", shelf="1.1", status="normal"} | 1
shelf_voltage_labels{cluster="fas8040-206-21", datacenter="DC-01", instance="localhost:12992", job="prometheus2", location="rear of the shelf on the lower right power supply", sensor_id="3", shelf="1.2", status="normal"} | 1
shelf_voltage_labels{cluster="fas8040-206-21", datacenter="DC-01", instance="localhost:12992", job="prometheus2", location="rear of the shelf on the lower right power supply", sensor_id="4", shelf="1.1", status="normal"} | 1
shelf_voltage_labels{cluster="fas8040-206-21", datacenter="DC-01", instance="localhost:12992", job="prometheus2", location="rear of the shelf on the lower right power supply", sensor_id="4", shelf="1.2", status="normal"}

### CODE REVIEW

**Review by cgrinds — DISMISSED** [review]
Nice job @Hardikl

**Comment by cgrinds** [comment]
This PR fixes #262

**Review by rahulguptajss — DISMISSED** [review]

**Comment by Hardikl** [comment]
shelf_status:

shelf_status{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", shelf="1.0"} | 0
-- | --
shelf_status{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", shelf="1.10"} | 0
shelf_status{cluster="fas8040-206-21", datacenter="DC-01", instance="localhost:12992", job="prometheus2", shelf="1.1"} | 0
shelf_status{cluster="fas8040-206-21", datacenter="DC-01", instance="localhost:12992", job="prometheus2", shelf="1.2"} | 0


shelf_labels:

shelf_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", model="DS2246", module_type="iom6", op_status="normal", serial_number="SHFFG1515000138", shelf="1.10", state="online", vendor_name="NETAPP "}
1
shelf_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", model="DS2246", module_type="iom6", op_status="normal", serial_number="SHFFG1524000057", shelf="1.0", state="online", vendor_name="NETAPP "}
1
shelf_labels{cluster="fas8040-206-21", datacenter="DC-01", instance="localhost:12992", job="prometheus2", model="DS2246", module_type="iom6", op_status="normal", serial_number="SHFMS1423000492", shelf="1.2", state="online", vendor_name="NETAPP"}
1
shelf_labels{cluster="fas8040-206-21", datacenter="DC-01", instance="localhost:12992", job="prometheus2", model="DS2246", module_type="iom6", op_status="normal", serial_number="SHFMS1423001152", shelf="1.1", state="online", vendor_name="NETAPP"}
1

**Comment by vgratian** [comment]
good job, do we still need the `state` label as well or can we remove it?

also, would be good when we update the dashboard to make use of the op-status of fans and sensors..

**Review by vgratian — APPROVED** [review]

**Comment by Hardikl** [comment]
Made the changes for child objects like psu, fans, etc.

when the op_status is normal, shelf_sensor_labels metric won't show records.
But when op_status is non normal (ex: unknown), shelf_sensor_labels metric showing full details like this:

shelf_sensor_labels{cluster="F8080-32-25", datacenter="DC-03", instance="localhost:12991", job="prometheus1", location="rear of the shelf on the lower left power supply", sensor_id="1", shelf="1.0", status="unknown"} | 1
-- | --

File: cmd/collectors/zapi/plugins/shelf/shelf.go [hunk 1]
```diff
@@ -197,13 +199,23 @@ func (my *Shelf) Run(data *matrix.Matrix) ([]*matrix.Matrix, error) {
...
+							}
 						}
 					}
 
 					instance.SetLabel("shelf", shelfName)
 					instance.SetLabel("shelf_id", shelfId)
-
+					// Remove instance from metric
+					if !isInclude {
+						data1.RemoveInstance(shelfId + "." + key)
```
Comment (cgrinds) [inline cmd/collectors/zapi/plugins/shelf/shelf.go:216]: minor: can you pull out `shelfId + "." + key` into a separate var on the line above and then use the new var on this line and the log line?

File: cmd/collectors/zapi/plugins/shelf/shelf.go [hunk 2]
```diff
@@ -184,6 +184,8 @@ func (my *Shelf) Run(data *matrix.Matrix) ([]*matrix.Matrix, error) {
 			my.Logger.Debug().Msgf("fetching %d [%s] instances", len(objectElem.GetChildren()), attribute)
 
 			for _, obj := range objectElem.GetChildren() {
+				// To decide instance would be available in metric or not
+				isInclude := true
```
Comment (cgrinds) [inline cmd/collectors/zapi/plugins/shelf/shelf.go:188]: minor: but can you move this down closer to where it's used, ~ line 199

File: cmd/collectors/zapi/plugins/shelf/shelf.go (hunk 1 above)
Comment (Hardikl) [reply cmd/collectors/zapi/plugins/shelf/shelf.go:216]: Done.

File: cmd/collectors/zapi/plugins/shelf/shelf.go (hunk 2 above)
Comment (Hardikl) [reply cmd/collectors/zapi/plugins/shelf/shelf.go:188]: done.

**Review by cgrinds — APPROVED** [review]

**Comment by hashi825** [comment]
Wouldn't it make sense to keep all instances for use with Grafana dashboards? 
Is it possible to just have shelf _labels metrics just be converted to status?

Example:

`shelf_fan_labels{cluster="cluster1", datacenter="dc1", fan_id="1", instance="localhost:12994", job="harvest", location="rear of the shelf on the lower left power supply", shelf="2.0",status="normal"} 1`

Would become

`shelf_fan_status{cluster="cluster1", datacenter="dc1", fan_id="1", instance="localhost:12994", job="harvest", location="rear of the shelf on the lower left power supply", shelf="2.0"}  1`

The value mapping of op_status for that child object would map to 1 for normal and other values for non normal, or would that require doing value mappings for each possible value? (which op-status values in the SDK differs in each shelf child object, fan,psu,sensors etc)

If it's not possible, I'd probably recommend keeping all child instances, this way you can use them in dashboards otherwise if healthy then they won't exist for visualisation.

**Comment by Hardikl** [comment]
Hi @hashi825 

your current suggestion is little different than previous one: expr: shelf_sensor_labels{status!="normal"}, where only non normal status would be expected.

Wouldn't it make sense to keep all instances for use with Grafana dashboards?
--> By default it was there before yesterday's change.

Let me evaluate the example you have mentioned for 'shelf_fan_status', and come back with my findings.

**Comment by Hardikl** [comment]
@hashi825 
I have modified yesterday's change along with new changes so that shelf_fan_status, like metrics would be available. #313 

Each child would have different possible values which is ugly way to write all of them and mapped manually, And as these are child of shelf object, so normal value would be mapped to 1 and rest all are mapped to 0 in newly added metric.

if op-status would be normal:
image

if op-status would be unknown(programmatically changed to unknown):
image

**Comment by hashi825** [comment]
That's great, with this you can just use shelf_fan_status = 0 for alerting which is much simpler and for grafana dashboards you should be able to merge with the shelf_fan_labels series to display full information in a table.

**Pull request merged**